### PR TITLE
feat: add EarnService to SDK with buildCreatePositionTx

### PR DESCRIPTION
**Pull request by 0xKoaj**
This pull request adds the EarnService to the SDK with the buildCreatePositionTx function. The EarnService allows users to create earn positions and manage permissions for those positions.

### CODE REVIEW

File: src/services/earn/types.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+export type IEarnService = {};
+
+export type CreateEarnPositionParams = {
+  chainId: ChainId;
+  strategyId: bigint;
+  depositToken: TokenAddress;
+  owner: Address;
+  permissions: EarnPermissionSet[];
+  strategyValidationData: Hex;
+  misc: Hex;
```
Comment (0xsambugs) [inline src/services/earn/types.ts:14]: Let's make these two optional, and default to an empty string when they are not set

File: src/services/earn/earn-service.ts [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+async function buildCompanionMulticall({ chainId, calls, value }: { chainId: ChainId; calls: Call[]; value?: bigint }) {
+  const data = encodeFunctionData({
+    abi: companionAbi,
+    functionName: 'multicall',
+    args: [calls],
+  });
+  return { to: EARN_VAULT_COMPANION.address(chainId), data, value };
+}
+
+type Call = Hex;
```
Comment (0xsambugs) [inline src/services/earn/earn-service.ts:206]: Let's use `Hex` directly

File: src/services/earn/types.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
+import { PermitData } from '@services/permit2';
+import { Address, BigIntish, ChainId, TimeString, TokenAddress } from '@types';
+import { Hex } from 'viem';
+
+export type IEarnService = {};
+
+export type CreateEarnPositionParams = {
+  chainId: ChainId;
+  strategyId: bigint;
+  depositToken: TokenAddress;
```
Comment (0xsambugs) [inline src/services/earn/types.ts:10]: Let's remove this like we discussed offline

**Review by 0xsambugs — COMMENTED** [review]
Great work!
Left some feedback

**Review by 0xsambugs — COMMENTED** [review]
Great work, just a few minor changes and we are done

File: src/services/earn/earn-service.ts [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+            owner as ViemAddress,
+            permissions.map(({ operator, permissions }) => ({
+              operator: operator as ViemAddress,
+              permissions: permissions.map(mapPermission),
+            })),
+            strategyValidationData ?? '0x',
+            misc ?? '0x',
+          ],
+        }),
+        value: isSameAddress(depositInfo.token, Addresses.NATIVE_TOKEN) ? depositInfo.amount : undefined,
```
Comment (0xsambugs) [inline src/services/earn/earn-service.ts:64]: I think you already have this here
```suggestion
        value: depositInfo.value,
```

File: src/services/earn/types.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
+import { PermitData } from '@services/permit2';
+import { Address, BigIntish, ChainId, TimeString, TokenAddress } from '@types';
+import { Hex } from 'viem';
+
+export type IEarnService = {};
```
Comment (0xsambugs) [inline src/services/earn/types.ts:5]: Could we add the function here?

File: src/services/earn/earn-service.ts [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+    if ('permitData' in deposit) {
+      calls.push(buildTakeFromCallerWithPermit(deposit.permitData, deposit.signature, recipient));
+    } else if (!isSameAddress(depositInfo.token, Addresses.NATIVE_TOKEN)) {
+      calls.push(buildTakeFromCaller(depositInfo.token, depositInfo.amount, recipient));
+    }
+
+    const depositToken = (needsSwap ? asset : depositInfo.token) as ViemAddress;
+
+    const [, allowances] = await Promise.all([
+      async () => {
```
Comment (0xsambugs) [inline src/services/earn/earn-service.ts:81]: This works? Wow. TIL

File: src/services/earn/earn-service.ts [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+import { IProviderService } from '@services/providers';
+import { MULTICALL_CONTRACT } from '@services/providers/utils';
+import { IAllowanceService } from '@services/allowances';
+
+export class EarnService implements IEarnService {
+  constructor(
+    private readonly permit2Service: IPermit2Service,
+    private readonly quoteService: IQuoteService,
+    private readonly providerService: IProviderService,
+    private readonly allowanceservice: IAllowanceService
```
Comment (0xsambugs) [inline src/services/earn/earn-service.ts:22]:
```suggestion
    private readonly allowanceService: IAllowanceService
```

File: src/services/earn/earn-service.ts [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+              buyToken: asset,
+              order: { type: 'sell', sellAmount: depositInfo.amount },
+            },
+            leftoverRecipient: owner,
+            swapConfig: deposit?.swapConfig,
+          });
+          calls.push(swapData);
+        }
+      },
+      this.allowanceservice.getAllowances({
```
Comment (0xsambugs) [inline src/services/earn/earn-service.ts:97]: You can use `getAllowanceInChain` which avoids the need for an array and record

File: src/services/earn/earn-service.ts [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+        allowances: [
+          {
+            chainId,
+            owner,
+            spender: EARN_VAULT_COMPANION.address(chainId),
```
Comment (0xsambugs) [inline src/services/earn/earn-service.ts:102]:
```suggestion
            owner: EARN_VAULT_COMPANION.address(chainId),
            spender: EARN_VAULT.address(chainId),
```

File: src/services/earn/earn-service.ts (hunk 8 above)
Comment (0xsambugs) [inline src/services/earn/earn-service.ts:97]: We should only check this if the deposit token is not the native address

File: src/services/earn/earn-service.ts [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+            token: depositToken,
+          },
+        ],
+      }),
+    ]);
+
+    // Only if the allowance is 0 we need to max approve the vault
+    const maxApprove = allowances[chainId]?.[owner]?.[depositToken]?.[EARN_VAULT_COMPANION.address(chainId)] != 0n;
```
Comment (0xsambugs) [inline src/services/earn/earn-service.ts:110]:
```suggestion
    const maxApprove = allowances[chainId]?.[owner]?.[depositToken]?.[EARN_VAULT_COMPANION.address(chainId)] === 0n;
```

**Review by 0xsambugs — COMMENTED** [review]
Some more comments

File: src/services/earn/earn-service.ts (hunk 6 above)
Comment (0xKoaj) [reply src/services/earn/earn-service.ts:81]: Not really, it needs one `()` at the end

File: src/services/earn/earn-service.ts [hunk 11]
```diff
@@ -0,0 +1,267 @@
...
+          },
+        ],
+      }).then(allowances => {
+        // Only if the allowance is 0 we need to max approve the vault
+        maxApprove = allowances[companion]?.[depositToken]?.[vault] === 0n;
+      });
+      promises.push(allowancePromise);
+    }
+    
+    await Promise.all(promises);
```
Comment (0xsambugs) [inline src/services/earn/earn-service.ts:115]: Like how you handled this!

File: src/services/earn/earn-service.ts [hunk 12]
```diff
@@ -0,0 +1,267 @@
...
+          order: { type: 'sell', sellAmount: depositInfo.amount },
+        },
+        leftoverRecipient: owner,
+        swapConfig: deposit?.swapConfig,
+      }).then(({ swapData }) => calls.push(swapData));
+      promises.push(swapPromise);
+    }
+    
+    if (!isSameAddress(depositToken, Addresses.NATIVE_TOKEN)) {
+      const allowancePromise = this.allowanceService.getAllowancesInChain({
```
Comment (0xsambugs) [inline src/services/earn/earn-service.ts:99]: Maybe we could use `getAllowanceInChain` instead? It's easier to use for our scenario

**Review by 0xsambugs — APPROVED** [review]
Great work!

**Pull request merged**